### PR TITLE
Improve Node.js `Readable`/`Writable` converters, introduce `Duplex` converter

### PR DESCRIPTION
**Pull request by armanbilge**
Idiomatic APIs as inspired by their [JVM cousins](https://github.com/typelevel/fs2/blob/main/io/jvm/src/main/scala/fs2/io/ioplatform.scala).

Edit: this PR is becoming a catch-all for other fixes.

Edit 2: okay, I ended up doing some major reworking in anticipation of new socket implementations.

### CODE REVIEW

File: io/js/src/main/scala/fs2/io/ioplatform.scala [hunk 1]
```diff
@@ -134,7 +137,12 @@ private[fs2] trait ioplatform {
         }.drain
       }
 
-  def mkWritable[F[_]](implicit F: Async[F]): Resource[F, (Writable, Stream[F, Byte])] =
+  def mkWritable[F[_]: Async](f: Writable => F[Unit]): Stream[F, Byte] =
```
Comment (armanbilge) [inline io/js/src/main/scala/fs2/io/ioplatform.scala:140]: Not sure about the name of this method, thoughts?

Comment (mpilquist) [reply io/js/src/main/scala/fs2/io/ioplatform.scala:140]: `readWritable`? Not my favorite but matches the convention used on JVM side

Comment (armanbilge) [reply io/js/src/main/scala/fs2/io/ioplatform.scala]: Not a fan myself, but I do like the consistency. Also renamed `from{Readable,Writable}` to `readReadable` and `writeWritable`.

File: io/js/src/main/scala/fs2/io/ioplatform.scala [hunk 2]
```diff
@@ -41,12 +41,14 @@ import scala.scalajs.js.|
...
+  ): Stream[F, Byte] =
     Stream
       .resource(for {
         readable <- Resource.makeCase(readable.map(_.asInstanceOf[streamMod.Readable])) {
           case (readable, Resource.ExitCase.Succeeded) =>
-            if (!readable.readableEnded)
+            if (!readable.readableEnded & destroyIfNotEnded)
               F.delay(readable.destroy())
             else
               F.unit
```
Comment (armanbilge) [inline io/js/src/main/scala/fs2/io/ioplatform.scala:54]: Still not thrilled with this but I think it is the best we can do. The underlying issue is that Node.js has no way to close/"end" a `Readable` besides the nuclear option `destroy`. This same thing came up in https://github.com/typelevel/fs2/pull/2453#discussion_r666153173 when we couldn't implement the `endOfInput` method for sockets.

File: io/js/src/main/scala/fs2/io/ioplatform.scala [hunk 3]
```diff
@@ -131,77 +119,109 @@ private[fs2] trait ioplatform {
...
         go(in).stream.handleErrorWith { ex =>
           Stream.eval(F.delay(writable.destroy(js.Error(ex.getMessage))))
         }.drain
       }
 
-  def mkWritable[F[_]](implicit F: Async[F]): Resource[F, (Writable, Stream[F, Byte])] =
+  def readWritable[F[_]: Async](f: Writable => F[Unit]): Stream[F, Byte] =
+    Stream.empty.through(toDuplexAndRead(f))
+
+  def toDuplexAndRead[F[_]: Async](f: Duplex => F[Unit]): Pipe[F, Byte, Byte] =
```
Comment (armanbilge) [inline io/js/src/main/scala/fs2/io/ioplatform.scala:136]: An even funkier name! Thoughts? 😅

**Comment by armanbilge** [comment]
@mpilquist apologies for all the churn. In these latest commits I introduced the `Duplex` (i.e., readable and writable) facade along with a converter. To avoid duplicating code, I reimplemented the readable/writable converters in terms of the duplex converter although this does add quite a bit of overhead. It is very unfortunate that Node.js doesn't offer a way to create a `Duplex` from a `Readable` and a `Writable`.

`Duplex` streams are pretty important in Node.js (e.g., their sockets implement `Duplex`) so the idea is to use these changes to cleanup the `Socket` implementation in a forthcoming PR.

Thanks!

**Comment by armanbilge** [comment]
Closing in favor of #2504 because it's easier to debug these converters alongside the socket implementations.

**Pull request closed without merging**